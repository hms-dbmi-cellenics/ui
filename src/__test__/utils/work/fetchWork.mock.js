const mockGenesListData = ['A', 'B', 'C', 'D'];

// const mockGeneExpressionData = {
//   A: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//   },
//   B: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.702857143,
//       stdev: 2.551115536,
//       expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.702857143,
//       stdev: 2.551115536,
//       expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
//     },
//   },
//   C: {
//     rawExpression: {
//       min: 0,
//       max: 3.4,
//       mean: 1.68,
//       stdev: 2.141525936,
//       expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 3.4,
//       mean: 1.68,
//       stdev: 2.141525936,
//       expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
//     },
//   },
//   D: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//   },
//   E: { hello: 'world' },
// };

// const mockCacheGet = jest.fn((x) => {
//   const mockCacheContents = {
//     fd3161a878f67ebf54018720cffd6a66: 'A', // pragma: allowlist secret
//     '10250a11679234110a1c260d6fd81d3c': 'B', // pragma: allowlist secret
//     f5c957411a28de68f35e1f5c8a29da7e: 'C', // pragma: allowlist secret
//     '33faa711a94a2028b5bae1778126aec0': 'E', // pragma: allowlist secret
//   };

//   if (x in mockCacheContents) {
//     return mockGeneExpressionData[mockCacheContents[x]];
//   }

//   return null;
// });

const mockCacheGet = jest.fn();
const mockCacheSet = jest.fn();
const mockCacheRemove = jest.fn();

const mockFetchWorkRequest = jest.fn(() => true);

const mockSeekFromS3 = jest.fn();

const mockCacheModule = {
  get: jest.fn((x) => mockCacheGet(x)),
  set: jest.fn((key, val) => mockCacheSet(key, val)),
  _remove: jest.fn((key) => mockCacheRemove(key)),
};

const mockFetchWorkModule = {
  __esModule: true,
  fetchWork: mockFetchWorkRequest,
};

const mockQcPipelineStartDate = '2021-01-01T01:01:01.000Z';

export {
  mockGenesListData, mockCacheGet, mockCacheSet,
  mockFetchWorkRequest, mockSeekFromS3,
  mockQcPipelineStartDate, mockCacheModule, mockFetchWorkModule,
};
