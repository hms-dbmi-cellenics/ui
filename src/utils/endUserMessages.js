export default {
  ACCOUNT_DETAILS_UPDATED: 'Account details updated successfully.',
  CONNECTION_ERROR: 'Could not connect to the server. Check your internet connection and refresh the page.',
  DELETING_PROJECT: 'Deleting project...',
  DELETING_SAMPLE: 'Deleting sample...',
  EMPTY_CLUSTER_NOT_CREATED: 'Cannot create a cluster that would be empty.',
  ERROR_DELETING_PROJECT: 'We couldn\'t delete your project.',
  ERROR_DELETING_SAMPLES: 'We couldn\'t delete samples from your project.',
  ERROR_DOWNLOADING_DATA: 'Error downloading data',
  ERROR_FETCHING_CELL_SETS: 'We couldn\'t get the list of cell sets.',
  ERROR_FETCHING_EXPERIMENTS: 'We couldn\'t get the list of experiments.',
  ERROR_FETCHING_PLOT_CONFIG: 'We couldn\'t get the plot settings.',
  ERROR_FETCHING_PLOT_DATA: 'We couldn\'t get the plot data.',
  ERROR_FETCHING_PROCESSING: 'We couldn\'t get the filter settings.',
  ERROR_FETCHING_PROJECTS: 'We couldn\'t get the list of projects.',
  ERROR_FETCHING_SAMPLES: 'We couln\'t get your samples.',
  ERROR_FETCHING_STATUS: 'We couldn\'t get the status of pending jobs.',
  ERROR_FETCHING_BACKEND_STATUS: 'We couldn\'t get the status of processing the experiment\'s processing run.',
  ERROR_NOT_SIGNED_IN: 'You need to sign in first. Click Sign in on the top right to begin.',
  ERROR_SAVING: 'We couldn\'t save your data.',
  ERROR_STARTING_PIPLELINE: 'We couldn\'t start working on your data.',
  ERROR_SIGN_IN: 'We couldn\'t sign you in. Please try again.',
  FEEDBACK_ERROR: 'We couldn\'t process your feedback. Please try submitting again.',
  FEEDBACK_SUCCESSFUL: 'Your feedback has been successfully sent. Thank you!',
  REFERRAL_ERROR: 'We couldn\'t process your referral. Please try submitting again.',
  REFERRAL_EMAIL_ERROR: 'The error you submitted was invalid. Please check again.',
  REFERRAL_SUCCESSFUL: 'Your invitation has been successfully sent. Thank you!',
  SUCCESS_NEW_CLUSTER_CREATED: 'Created a new cluster. Check the `Data Management` window.',
  SUCCESS_CELL_SETS_RECLUSTERED: 'Cell sets have been reclustered. Your cell sets have changed.',
  NO_CLUSTERS: 'No custom cell sets available, to create more go to data-exploration.',
  SAVING_SAMPLE: 'Saving sample...',
  SAVING_PROJECT: 'Saving project...',
};
