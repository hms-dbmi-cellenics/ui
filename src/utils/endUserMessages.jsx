export default {
  ACCOUNT_DETAILS_UPDATED: 'Account details updated successfully.',
  CONNECTION_ERROR: 'Could not connect to the server. Check your internet connection and refresh the page.',
  DELETING_PROJECT: 'Deleting project...',
  DELETING_SAMPLE: 'Deleting sample...',
  ERROR_CREATING_SAMPLE: 'We couldn\'t create your sample. Please try uploading it again',
  ERROR_CREATING_PROJECT: 'We couldn\'t create your project.',
  ERROR_LOADING_PROJECT: 'We couldn\'t load your project.',
  EMPTY_CLUSTER_NOT_CREATED: 'Cannot create a cluster that would be empty.',
  ERROR_DELETING_PROJECT: 'We couldn\'t delete your project.',
  ERROR_DELETING_SAMPLES: 'We couldn\'t delete samples from your project.',
  ERROR_DOWNLOADING_DATA: 'Error downloading data.',
  ERROR_FETCHING_CELL_SETS: 'We couldn\'t get the list of cell sets.',
  ERROR_FETCHING_EXPERIMENTS: 'We couldn\'t get the list of experiments.',
  ERROR_FETCHING_PLOT_CONFIG: 'We couldn\'t get the plot settings.',
  ERROR_FETCHING_PLOT_DATA: 'We couldn\'t get the plot data.',
  ERROR_FETCHING_NORMALIZED_EXPRESSION_MATRIX: 'We couldn\'t download the normalized expression matrix.',
  ERROR_NO_MATCHING_CELLS_NORMALIZED_EXPRESSION_MATRIX: 'No cells match requested filters.',
  ERROR_FETCHING_PROCESSING: 'We couldn\'t get the filter settings.',
  ERROR_FETCHING_DATA_PROCESSING_SETTINGS: 'We couldn\'t get the data processing settings.',
  ERROR_FETCHING_PROJECTS: 'We couldn\'t get the list of projects.',
  ERROR_FETCHING_SAMPLES: 'We couln\'t get your samples.',
  ERROR_FETCHING_STATUS: 'We couldn\'t get the status of pending jobs.',
  ERROR_FETCHING_BACKEND_STATUS: 'We couldn\'t get the status of the experiment\'s data processing.',
  ERROR_FETCH_BACKGROUND_GENE_EXP: 'We couldn\'t get the background gene expression',
  ERROR_FETCH_MARKER_GENES: 'We couldn\'t get the marker genes',
  ERROR_FILE_CATEGORY: 'The selected file name does not match the expected category.',
  ERROR_FILES_FOLDER: 'Only files contained in a folder are accepted.',
  ERROR_METADATA_MULTIPLE_FILES: 'You need to upload a single file in .tsv format',
  ERROR_METADATA_WRONG_FORMAT: 'The file does not seem to be in the correct tsv key-value format.',
  ERROR_CELL_LEVEL_COLUMN: 'The .tsv file needs to contain the column "barcode"',
  ERROR_OBJ2S_MULTIPLE_FILES: 'Only a single file is accepted for uploads.',
  ERROR_OBJ2S_EXISTING_FILE: 'Please remove existing file before uploading a new one.',
  ERROR_OBJ2S_MAX_FILE_SIZE: 'file size must be less than 15 GB.',
  ERROR_NOT_SIGNED_IN: 'You need to sign in first. Click Sign in on the top right to begin.',
  ERROR_NO_PERMISSIONS: 'You do not have permissions to perform that action.',
  ERROR_LAUNCH_PATHWAY: 'We couldn\'t launch pathway analysis',
  ERROR_SAVING: 'We couldn\'t save your data.',
  ERROR_BEGIN_SAMPLE_FILE_UPLOAD: 'We couldn\'t begin the files upload',
  ERROR_UPDATE_SERVER_ON_UPLOAD_STATE: 'We couldn\'t update the upload state in our servers. Try refreshing the page.',
  ERROR_SAMPLE_TECHNOLOGY: 'Selected technology does not match across all samples. Select the right technology and try again.',
  ERROR_SAVING_PLOT_CONFIG: 'We couldn\'t save your plot configuration.',
  ERROR_STARTING_PIPLELINE: 'We couldn\'t start working on your data.',
  ERROR_SIGN_IN: 'We couldn\'t sign you in. Please try again.',
  ERROR_DOWNLOADING_OBJ2S: 'We couldn\'t get the processed object.',
  FEEDBACK_ERROR: 'We couldn\'t process your feedback. Please try submitting again.',
  FEEDBACK_SUCCESSFUL: 'Your feedback has been successfully sent. Thank you!',
  REFERRAL_ERROR: 'We couldn\'t process your referral. Please try submitting again.',
  REFERRAL_EMAIL_ERROR: 'The error you submitted was invalid. Please check again.',
  REFERRAL_SUCCESSFUL: 'Your invitation has been successfully sent. Thank you!',
  SUCCESS_NEW_CLUSTER_CREATED: (
    <>
      Created a new cluster. Check &nbsp;
      <b>Custom cell sets</b>
      &nbsp; in the &nbsp;
      <b>Cell sets and Metadata </b>
      tile.
    </>),
  SUCCESS_CELL_SETS_RECLUSTERED: 'Cell sets have been reclustered. Your cell sets have changed.',
  SUCCESS_CELL_SETS_ANNOTATED: 'Cell sets have been annotated. Your cell sets have changed.',
  WORK_REQUEST_TIMED_OUT_RETRY: 'The task is taking longer than expected, please retry.',
  SAVING_SAMPLE: 'Saving sample...',
  SAVING_PROJECT: 'Saving project...',
  SHARE_SUCESS_NOTIFICATION_FAILURE: 'The project was shared, but we werenâ€™t able to notify the new collaborator',
  SHARE_FAILURE: 'We weren\'t able to share the project',
  ERROR_CLONING_DEFAULT: 'We couldn\'t clone the project.',
  ERROR_CLONING_PIPELINE_LOCKED: 'We couldn\'t clone the project because data processing is currently running, wait until it finishes and try again',
  ERROR_CELL_SETS_ANNOTATION_FAILED: 'ScType annotation unsuccessful. Please ensure dataset contains gene symbols.',
};
