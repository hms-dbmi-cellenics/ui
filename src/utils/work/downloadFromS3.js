import unpackResult from 'utils/work/unpackResult';
import httpStatusCodes from 'utils/http/httpStatusCodes';
import parseResult from 'utils/work/parseResult';

const downloadFromS3 = async (taskName, signedUrl) => {
  const response = await fetch(signedUrl);

  // some WorkRequests like scType and runClustering do not upload data to S3
  // (nor return it via socket) instead they patch the cellsets through the API.
  // In those cases the workResults will not exist, and it's fine because data will
  // be updated through the cellsets.
  // the forbidden (in addition to the not found) is required because of how signed URLs work
  //  when you try to download a file from a signedUrl that doesn't exist you get a 403 forbidden
  // because the user is not authorized to access a file that does not exist
  if (response.status === httpStatusCodes.NOT_FOUND
    || response.status === httpStatusCodes.FORBIDDEN) {
    return null;
  }
  if (!response.ok) {
    throw new Error(`Error ${response.status}: ${response.text}`, { cause: response });
  }

  const unpackedResult = await unpackResult(response, taskName);
  const parsedResult = parseResult(unpackedResult, taskName);

  return parsedResult;
};

export default downloadFromS3;
