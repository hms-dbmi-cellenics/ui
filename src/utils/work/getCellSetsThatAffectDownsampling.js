import loadCellSets from 'redux/actions/cellSets/loadCellSets';
import { getCellSetsHierarchy, getCellSetsHierarchyByKeys } from 'redux/selectors';

// Check that the cell sets within the selected selectedCellSet and grouped tracks didn't change
// e.g., if cell set was deleted we can't use cache
const getCellSetsThatAffectDownsampling = async (
  experimentId, selectedCellSetKey, groupedTracks, dispatch, getState) => {
  await dispatch(loadCellSets(experimentId));

  const [{ children }] = getCellSetsHierarchyByKeys([selectedCellSetKey])(getState());

  const selectedCellSetsKeys = children.map((cellSet) => cellSet.key);

  const groupedCellSetKeys = getCellSetsHierarchy(groupedTracks)(getState())
    .map((cellClass) => cellClass?.children)
    .flat()
    .map(({ key }) => key);

  // Keep them in separate lists, they each represent different changes in the settings
  return [selectedCellSetsKeys, groupedCellSetKeys];
};

export default getCellSetsThatAffectDownsampling;
