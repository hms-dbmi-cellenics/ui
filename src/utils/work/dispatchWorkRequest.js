import dayjs from 'dayjs';
import { Environment } from 'utils/deploymentInfo';

import fetchAPI from 'utils/http/fetchAPI';
// Disable unique keys to reallow reuse of work results in development
const DISABLE_UNIQUE_KEYS = [
  'GetEmbedding',
];

const getCacheUniquenessKey = (taskName, environment) => {
  // Disable cache in development or if localStorage says so
  // Do not disable for embeddings requests because download seurat & trajectory depend on that ETag
  if (
    environment !== Environment.PRODUCTION
    && (localStorage.getItem('disableCache') === 'true' || environment === Environment.DEVELOPMENT)
    && !DISABLE_UNIQUE_KEYS.includes(taskName)
  ) {
    return Math.random();
  }

  return null;
};

const getWorkerTimeout = (taskName, defaultTimeout) => {
  switch (taskName) {
    case 'GetEmbedding':
    case 'ListGenes':
    case 'MarkerHeatmap': {
      return dayjs().add(1800, 's').toISOString();
    }

    default: {
      return dayjs().add(defaultTimeout, 's').toISOString();
    }
  }
};

const dispatchWorkRequest = async (
  experimentId,
  body,
  timeout,
  requestProps,
  getState,
) => {
  const { name: taskName } = body;

  // for listGenes, markerHeatmap, & getEmbedding we set a long timeout for the worker
  // after that timeout the worker will skip those requests
  // meanwhile in the UI we set a shorter timeout. The UI will be prolonging this timeout
  // as long as it receives "heartbeats" from the worker because that means the worker is alive
  // and progresing.
  // this should be removed if we make each request run in a different worker
  const workerTimeoutDate = getWorkerTimeout(taskName, timeout);

  const { environment } = getState().networkResources;
  const cacheUniquenessKey = getCacheUniquenessKey(taskName, environment);

  // eslint-disable-next-line no-param-reassign
  requestProps.cacheUniquenessKey = cacheUniquenessKey;

  const request = {
    experimentId,
    timeout: workerTimeoutDate,
    body,
    requestProps,

  };

  const { data: { ETag, signedUrl } } = await fetchAPI(
    `/v2/workRequest/${experimentId}`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    },
  );

  return { ETag, signedUrl, request };
};

export default dispatchWorkRequest;
