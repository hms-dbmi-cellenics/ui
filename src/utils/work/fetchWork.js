import { Environment, isBrowser } from 'utils/deploymentInfo';

import cache from 'utils/cache';
import dispatchWorkRequest from 'utils/work/dispatchWorkRequest';
import downloadFromS3 from 'utils/work/downloadFromS3';
import waitForWorkRequest from 'utils/work/waitForWorkRequest';

// retrieveData will try to get the data for the given experimentId and ETag from
// the fastest source possible. It will try to get the data in order from:
// 1. Browser cache
// 2. S3 (cache) via signedURL
// 3. Worker
//   3.1. Via socket (small data)
//   3.2. Via S3 (large data) via signedURL
const retrieveData = async (experimentId,
  ETag,
  signedUrl,
  request,
  timeout,
  body,
  dispatch,
  useCache) => {
  // 1. Check if we have the ETag in the browser cache (no worker)
  const cachedData = useCache ? await cache.get(ETag) : null;
  if (cachedData) {
    return cachedData;
  }

  // 2. Check if data is cached in S3 so we can download from the signed URL (no worker)
  if (signedUrl) {
    return await downloadFromS3(body.name, signedUrl);
  }

  // 3. If we don't have signedURL, wait for the worker to send us the data via
  // - the data via socket
  // - the signedURL to download the data from S3
  const { signedUrl: workerSignedUrl, data } = await waitForWorkRequest(
    ETag,
    experimentId,
    request,
    timeout,
    dispatch,
  );

  // 3.1. The worker send the data via socket because it's small enough
  if (data) {
    return data;
  }
  // 3.2. The worker send a signedUrl to download the data
  return await downloadFromS3(body.name, workerSignedUrl);
};

const fetchWork = async (
  experimentId,
  body,
  getState,
  dispatch,
  optionals = {},
) => {
  const {
    extras = undefined,
    timeout = 180,
    broadcast = false,
    onETagGenerated = () => { },
  } = optionals;

  if (!isBrowser) {
    throw new Error('Disabling network interaction on server');
  }

  const { environment } = getState().networkResources;
  if (environment === Environment.DEVELOPMENT && !localStorage.getItem('disableCache')) {
    localStorage.setItem('disableCache', 'true');
  }

  // 1. Contact the API to get ETag and possible S3 signedURL
  const { ETag, signedUrl, request } = await dispatchWorkRequest(
    experimentId,
    body,
    timeout,
    {
      broadcast,
      ...extras,
    },
    dispatch,
  );

  onETagGenerated(ETag);

  const useCache = body.name !== 'GeneExpression';

  // 2. Try to get the data from the fastest source possible
  const data = await retrieveData(
    experimentId, ETag, signedUrl, request, timeout, body, dispatch, useCache,
  );

  // 3. Cache the data in the browser
  if (useCache) {
    await cache.set(ETag, data);
  }

  return data;
};

export default fetchWork;
