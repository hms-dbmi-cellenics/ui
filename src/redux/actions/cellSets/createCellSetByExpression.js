import { fetchWork } from 'utils/work/fetchWork';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import endUserMessages from 'utils/endUserMessages';

const createCellSetByExpression = (
  experimentId, selectedGenes,
) => async (dispatch, getState) => {
  const body = {
    name: 'GetExpressionCellSets',
    genesConfig: selectedGenes,
  };

  try {
    const result = await fetchWork(experimentId, body, getState, { broadcast: true });
    return result;
  } catch (e) {
    console.error(e);
    pushNotificationMessage('error', endUserMessages.ERROR_FETCHING_CELL_SETS);
    throw new Error(e);
  }
};

export default createCellSetByExpression;
