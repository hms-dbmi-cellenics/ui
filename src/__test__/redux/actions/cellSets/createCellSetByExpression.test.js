import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import createCellSetByExpression from 'redux/actions/cellSets/createCellSetByExpression';

import fake from '__test__/test-utils/constants';

import { fetchWork } from 'utils/work/fetchWork';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import endUserMessages from 'utils/endUserMessages';
import { waitFor } from '@testing-library/dom';

jest.mock('utils/work/fetchWork', () => ({
  fetchWork: jest.fn(),
}));
jest.mock('utils/pushNotificationMessage');

const mockStore = configureMockStore([thunk]);
const store = mockStore();
const experimentId = fake.EXPERIMENT_ID;

const mockData = [
  {
    geneName: 'GeneA',
    comparisonType: 'greaterThan',
    thresholdValue: '0.00',
  },
  {
    geneName: 'GeneB',
    comparisonType: 'lessThan',
    thresholdValue: '0.05',
  },
];

describe('createCellSetByExpression', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Dispatces the correct body', async () => {
    await store.dispatch(createCellSetByExpression(experimentId, mockData));

    const params = fetchWork.mock.calls[0];
    expect(params).toMatchSnapshot();
  });

  it('Throws notification on error', async () => {
    fetchWork.mockImplementationOnce(() => Promise.reject(new Error('some error')));

    expect(async () => {
      await store.dispatch(createCellSetByExpression(experimentId, mockData));
    }).rejects.toThrow();

    waitFor(() => {
      expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
      expect(pushNotificationMessage).toHaveBeenCalledWith('error', endUserMessages.ERROR_FETCHING_CELL_SETS);
    });
  });
});
