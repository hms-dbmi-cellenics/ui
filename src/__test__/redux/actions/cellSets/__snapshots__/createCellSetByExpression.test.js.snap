// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCellSetByExpression Dispatces the correct body 1`] = `
Array [
  "testae48e318dab9a1bd0bexperiment",
  Object {
    "genesConfig": Array [
      Object {
        "comparisonType": "greaterThan",
        "geneName": "GeneA",
        "thresholdValue": "0.00",
      },
      Object {
        "comparisonType": "lessThan",
        "geneName": "GeneB",
        "thresholdValue": "0.05",
      },
    ],
    "name": "GetExpressionCellSets",
  },
  [Function],
  Object {
    "broadcast": true,
  },
]
`;
